### PR TITLE
feature: driver znb20 update

### PR DESCRIPTION
**Pull request by nataliejpg**
Changes proposed in this pull request:
- adds spectroscopy options to the ZNB driver
- cleans it up somewhat, adding better docstrings (maybe overkill...)
- extension to example notebook

@WilliamHPNielsen @giulioungaretti @alexcjohnson (not sure who best to mention...)

- [ ] what should it be called given that is can be used with ZNB20 and ZNB8?
- [x] does it handle 'on' and 'off' of spec mode properly
- [ ] do comments and docstrings make sense

### CODE REVIEW

File: qcodes/instrument_drivers/rohde_schwarz/ZNB20.py [hunk 1]
```diff
@@ -1,20 +1,23 @@
...
+    Hardware controlled parameter class for Rohde Schwarz RSZNB20/8 trace.
 
     Instrument returns an list of transmission data in the form of a list of
     complex numbers taken from a frequency sweep.
 
-    TODO:
-      - ability to choose for abs or db in magnitude return
+    TODO(nataliejpg): ability to choose for abs or db in magnitude return
     """
+
```
Comment (giulioungaretti) [inline qcodes/instrument_drivers/rohde_schwarz/ZNB20.py]: :unicorn:  missing class docstring, possibly obiovus but start stop npts are not super obvious from the outside

File: qcodes/instrument_drivers/rohde_schwarz/ZNB20.py [hunk 2]
```diff
@@ -141,14 +186,60 @@ def _set_npts(self, val):
...
+        elif val == 0:
+            self.write('SOUR:FREQ1:CONV:ARB:IFR 1, 1, 0, SWE')
+            self.write('SOUR:FREQ2:CONV:ARB:IFR 1, 1, 0, SWE')
+            self.write('SOUR:POW1:OFFS 0, CPAD')
+            self.write('SOUR:POW2:OFFS 0, CPAD')
+            self.write('SOUR:POW1:PERM 0')
+            self.write('SOUR:POW3:PERM 0')
+            self.write('SOUR:POW3:STAT 0')
+        else:
+            logging.error('cannot set mode %s' % val)
```
Comment (giulioungaretti) [inline qcodes/instrument_drivers/rohde_schwarz/ZNB20.py]: Either print or logging, with preference for logging <3

File: qcodes/instrument_drivers/rohde_schwarz/ZNB20.py [hunk 3]
```diff
@@ -141,14 +186,60 @@ def _set_npts(self, val):
...
-        self.write('SENS1:SWE:TYPE LIN')
-        self.write('SENS1:SWE:TIME:AUTO ON')
-        self.write('TRIG1:SEQ:SOUR IMM')
-        self.write('SENS1:AVER:STAT ON')
-        self.update_display_on()
-        self.start(1e6)
-        self.stop(2e6)
-        self.npts(10)
-        self.power(-50)
+    # Spectrocopy mode settings
```
Comment (giulioungaretti) [inline qcodes/instrument_drivers/rohde_schwarz/ZNB20.py]: tpyo ?

File: qcodes/instrument_drivers/rohde_schwarz/ZNB20.py [hunk 4]
```diff
@@ -126,6 +156,21 @@ def __init__(self, name, address, **kwargs):
         self.initialise()
         self.connect_message()
 
+    def initialise(self):
+        """
+        Initialise instrument to sweep linearly, automatically calculate and
+        sweep in minimum time, not require a trigger, do averaging if avg set.
+
+        Spectrocopy mode is set to be off
```
Comment (giulioungaretti) [inline qcodes/instrument_drivers/rohde_schwarz/ZNB20.py]: typo ?

File: qcodes/instrument_drivers/rohde_schwarz/ZNB20.py (hunk 2 above)
Comment (nataliejpg) [inline qcodes/instrument_drivers/rohde_schwarz/ZNB20.py]: #guerilla-debugging ;)

**Comment by nataliejpg** [comment]
anything else?

**Comment by giulioungaretti** [comment]
nope! Will look and most likely merge again tomrrow!
Thanks for the contribution!

File: qcodes/instrument_drivers/rohde_schwarz/ZNB20.py [hunk 5]
```diff
@@ -1,20 +1,38 @@
...
-    TODO:
-      - ability to choose for abs or db in magnitude return
+    Args:
+      name: parameter name
+      instrument: instrument the parameter belongs to
+      start: starting frequency of sweep
+      stop: ending frequency of sweep
+      npts: numper of points in frequency sweep
+
+     Methods:
```
Comment (jenshnielsen) [inline qcodes/instrument_drivers/rohde_schwarz/ZNB20.py]: Methods is indented one space too much which breaks rendering of the docs

File: .gitignore [hunk 6]
```diff
@@ -77,4 +77,6 @@ target/
 *.py~
 tmp/
 
+Other
```
Comment (jenshnielsen) [inline .gitignore]: Should probably be `# Other`

**Comment by jenshnielsen** [comment]
@nataliejpg Is there any reason not to merge this? I have left 2 minor comments but otherwise I think this is ready

**Comment by nataliejpg** [comment]
@jenshnielsen To be honest looking at it I'm not sure if it really adds to the driver because it is such a specific extension to it and really should be made quite a lot more general which would probably take a bit of time. I'm not advocating adding everything in the manual but what I have added here is super specific and I think it does more harm than good in general. If there is time/inclination I am very happy to be part of the effort to make a more general pr but if there isn't I might instead say close this one rather than merge it.

**Comment by jenshnielsen** [comment]
Ok, I need to make some changes to the driver for the use on T10. I might just copy the more general thinks from here then and leave the specific things out for now

**Comment by nataliejpg** [comment]
@jenshnielsen ok, I wasn't aware anyone was using the vna for more complex stuff. Maybe a 'clean' intermediary pr is the solution as you suggest.

**Comment by jenshnielsen** [comment]
They are not using it yet but they want some more "simple" features that this currently driver S12, S22, autoscale and so on

**Comment by nataliejpg** [comment]
@jenshnielsen I think that it would be pretty fast to implement having the wave ratio (ie s21, s22, a1/b2) etc as a parameter (and have options available depending on if you have a 2 port or 4 port instrument). Autoscale is just a function which should also be super easy. Then 'next generation' would then deal with external generators. This current version works for an external generator in one specific setup but has no flexibility in wave ratios which is like introducing something complicated without putting in the easy stuff first. Plus with wave ratios as a parameter I imagine you will address 95% of user needs

**Comment by nataliejpg** [comment]
@jenshnielsen but just that I don't have time to test it right now on a real instrument and from experience it's something that really needs testing so I don't want to edit it to look better but then not be able to test it and I can't really commit the time to it.

**Comment by jenshnielsen** [comment]
No problem, I don't expect you to do anything. I was just trying to make sure that none of your work was lost in case I make other changes. Thanks for the input

**Comment by jenshnielsen** [comment]
replaced by #618

**Pull request closed without merging**